### PR TITLE
Adding client tests

### PR DESCRIPTION
**Pull request by donny-dont**
This adds a `dart:html` and `dart:io` compatible `Client` test. A server is added for the clients to call.

### CODE REVIEW

**Comment by donny-dont** [comment]
@nex3 this adds the tests but some tests fail.

On the `dart:io` side
* Doesn't appear that any bytes are getting transferred over as `body` is empty when retrieved
* The `content-length` header isn't getting set

On the `dart:html` side
* The `content-length` is set automatically by the browser, but it also appends `content-length` to methods other than `GET` and `HEAD`

For both
* `content-type` is not getting set.
* We left out the `x-www-form-urlencoded` implementation

Anyways I'd like to land this and fix the issues individually.

@kevmoo that means that tests will be failing until the above is resolved.

File: test/client_test.dart [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+  group('client', () {
+    // The server url of the spawned server
+    var serverUrl;
+
+    setUp(() async {
+      var channel = spawnHybridUri('hybrid/server.dart');
+      serverUrl = Uri.parse(await channel.stream.first);
+    });
+
+    // HEAD
```
Comment (nex3) [inline test/client_test.dart]: These headers seem redundant with the groups.

File: test/client_test.dart [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+            'headers': {
+              'user-agent': [userAgent()],
+              'x-random-header': ['Value'],
+              'x-other-header': ['Other Value']
+            }
+          })));
+    });
+
+    test('readBytes throws an error for a 4** status code', () {
+      platformClient().readBytes(serverUrl.resolve('/error')).then((_) {
```
Comment (nex3) [inline test/client_test.dart]: Use async/await.

File: test/hybrid/client.dart [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
```
Comment (nex3) [inline test/hybrid/client.dart]: I don't understand why this is under a "hybrid" directory.

File: test/hybrid/server.dart [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  'connection',
+  'origin',
+  'referer',
+
+  // Dart IO headers
+  'cookie',
+  'host',
+];
+
+hybridMain(StreamChannel channel) async {
```
Comment (nex3) [inline test/hybrid/server.dart]: Document this. In particular, mention which tests it's used in and what the server does.

File: test/client_test.dart (hunk 2 above)
Comment (donny-dont) [inline test/client_test.dart]: Is there a specific way I should be using one of the `expect` functions? I wasn't able to get it to work otherwise.

Comment (nex3) [inline test/client_test.dart]: Looks like you figured it out.

**Pull request merged**